.emacs.d/lisp/package/packup.el

```diff
@@ -9,17 +9,197 @@
+  (cond
+   ((use-region-p)
+    (save-excursion
+      (let ((beg (region-beginning))
+            (end (region-end)))
+        (kotct/packup-mark-packages-in-region
+         (progn (goto-char beg) (line-beginning-position))
+         (progn (goto-char end) (line-beginning-position))))))
+   (t
+    (let ((inhibit-read-only t))
+      (kotct/packup-repeat-over-lines
+       (prefix-numeric-value arg)
+       (lambda ()
+         (forward-char 1)
+         (delete-char 1)
+         (insert kotct/packup-marker-char)))))))
+
+(defun kotct/packup-unmark (arg &optional interactive)
+  "Unmark the package for update at point in the Packup buffer.
+If a region is selected, unmark all the packages in the region.
+If an prefix-arg is passed unmark ARG times."
+  (interactive (list current-prefix-arg))
+  (let ((kotct/packup-marker-char ?\040))
+    (kotct/packup-mark arg interactive)))
+
+(defun kotct/packup-mark-all ()
+  (interactive)
+  (save-excursion
+    (kotct/packup-mark-packages-in-region (point-min) (point-max))))
+
+
+(defun kotct/packup-unmark-all ()
+    (interactive)
+    (save-excursion
+      (let ((kotct/packup-marker-char ?\040))
+        (kotct/packup-mark-packages-in-region (point-min) (point-max)))))
+
+(defun kotct/packup-initialize-buffer-contents ()
+  (package-refresh-contents)
+  (let ((install-list nil))
+    (dolist (package kotct/dependency-list)
+      (if (or (not (package-installed-p package))
+              (not (kotct/package-up-to-date-p package)))
+          (apply 'kotct/packup-insert-package-row (list package (package-desc-version (cadr (assq package package-alist)))))))))
+
+(defun kotct/packup-initialize-buffer ()
+  "Initializes the packup buffer."
+  (kill-all-local-variables)
+  (use-local-map packup-mode-map) ;; mapppppps
+  (setq major-mode 'packup-mode
+        mode-name "Packup"
+        buffer-read-only t ;; read only
+        mode-line-buffer-identification "packup")
+  (kotct/packup-initialize-buffer-contents))
+
+(defun kotct/packup-refresh ()
+  "Refreshes packages in current buffer."
+  (interactive)
+  (if (eq major-mode 'packup-mode)
+    (let ((inhibit-read-only t))
+      (erase-buffer)
+      (kotct/packup-initialize-buffer-contents))))
+
+(defun kotct/packup-help ()
+  (interactive)
+  (message "g-refresh m-mark u-unmark x-execute ?-help"))
+
+(defvar packup-mode-map
+  (let ((map (make-keymap)))
+    (define-key map "m" 'kotct/packup-mark)
+    (define-key map "M" 'kotct/packup-mark-all)
+    (define-key map "u" 'kotct/packup-unmark)
+    (define-key map "U" 'kotct/packup-unmark-all)
+    (define-key map "x" 'kotct/packup-do-update)
+    (define-key map "g" 'kotct/packup-refresh)
+    (define-key map "?" 'kotct/packup-help)
+    map))
+
+;;;###autoload
+(defun kotct/packup ()
+  "Creates an interactive buffer to install/update packages."
+  (interactive)
+  (let ((old-buf (current-buffer))
+        (buffer (create-file-buffer "*packup*"))) ;; this is wrong...
```

Comment:
Hmm, buffer has name `packup` for me with no cool asterisks.  We should figure out how.  Here is `list-packages` from `package.el` in Emacs, which seems to use `get-buffer-create` instead.

```el
;;;###autoload
(defun list-packages (&optional no-fetch)
  "Display a list of packages.
This first fetches the updated list of packages before
displaying, unless a prefix argument NO-FETCH is specified.
The list is displayed in a buffer named `*Packages*'."
  (interactive "P")
  (require 'finder-inf nil t)
  ;; Initialize the package system if necessary.
  (unless package--initialized
    (package-initialize t))
  ;; Integrate the package-menu with updating the archives.
  (add-hook 'package--post-download-archives-hook
            #'package-menu--post-refresh)
  (add-hook 'package--post-download-archives-hook
            #'package-menu--mark-or-notify-upgrades 'append)

  ;; Generate the Package Menu.
  (let ((buf (get-buffer-create "*Packages*")))
    (with-current-buffer buf
      (package-menu-mode)

      ;; Fetch the remote list of packages.
      (unless no-fetch (package-menu-refresh))

      ;; If we're not async, this would be redundant.
      (when package-menu-async
        (package-menu--generate nil t)))
    ;; The package menu buffer has keybindings.  If the user types
    ;; `M-x list-packages', that suggests it should become current.
    (switch-to-buffer buf)))
```

Reply:
uhm you're straight tripping if you run `M-x kotct/packup` it creates a buffer that is called `*packup*`. i'm like sure of that

i used `create-file-buffer` because it was what `dired` uses. i agree i should have used `get-buffer-create`. pushing that up now.

Reply:
On Emacs HEAD it doesn't have asterisks, then, or they aren't displayed properly.

Reply:
¯\_(ツ)_/¯